### PR TITLE
ssim, ssim_loss, ssim_loss_fast

### PR DESCRIPTION
**Pull request by nikopj**
Implements structural similarity index measurement (SSIM) loss, addressing FluxML/Flux.jl#2165. 
Main function is `ssim`, which is used by `ssim_loss` and `ssim_loss_fast`. The fast version uses a smaller averaging kernel instead of the standard 11x11 gaussian kernel.

1D/2D/3D inputs (3D/4D/5D channel+batch tensors) and GPU are accounted for. The corresponding kernel is by default a 1D/2D/3D gaussian kernel (see `ssim_kernel`). 

`ssim` is tested against ImageQualityIndexes's `assess_ssim` for grayscale and RGB images.

This doesn't have the full IQI capabilities (such as alpha, beta, gamma params), but it seems those are kept as default 1 in other libraries that implement ssim loss (ex [https://github.com/VainF/pytorch-msssim](https://github.com/VainF/pytorch-msssim)). I believe we would require that if we extended the implementation to have a MS-SSIM loss.

Feedback is greatly appreciated.

### PR Checklist

- [x] Tests are added
- [x] Entry in NEWS.md
- [x] Documentation, if applicable

### CODE REVIEW

File: test/cuda/losses.jl [hunk 1]
```diff
@@ -26,13 +25,32 @@ y = [1  0  0  0  1
...
-  y = rand(Float32, 3,3)
+    x = rand(Float32, 3,3)
+    y = rand(Float32, 3,3)
 
-  for loss in ALL_LOSSES
-    gpu_autodiff_test(loss, x, y)
-  end
+    for loss in ALL_LOSSES
+        gpu_autodiff_test(loss, x, y)
+    end
```
Comment (CarloLucibello) [inline test/cuda/losses.jl:53]: please stick to flux's indentation style in old and new

File: src/losses/utils.jl [hunk 2]
```diff
@@ -36,3 +36,39 @@ end
...
+    elseif N-2 == 3
+        ks = length(SSIM_KERNEL)
+        kernel = reshape(SSIM_KERNEL*SSIM_KERNEL', 1, ks, ks).*SSIM_KERNEL
+    else
+        throw("SSIM is only implemented for 3D/4D/5D inputs, dimension=$N provided.")
+    end
+    return reshape(T.(kernel), size(kernel)..., 1, 1)
+end
+ssim_kernel(x::Array{T, N}) where {T, N} = ssim_kernel(T, N)
+ssim_kernel(x::AnyCuArray{T, N}) where {T, N} = cu(ssim_kernel(T, N))
```
Comment (CarloLucibello) [inline src/losses/utils.jl:73]: since these functions are not of general utility they better go close to the ssim the losses

File: Project.toml [hunk 3]
```diff
@@ -48,6 +48,9 @@ FillArrays = "1a297f60-69ca-5386-bcde-b61e274b549b"
 IterTools = "c8e1da08-722c-5040-9ed9-7db0dc04731e"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
+Images = "916415d5-f1e6-5110-898d-aaa5f9f070e0"
+ImageQualityIndexes = "2996bd0c-7a13-11e9-2da2-2f5ce47296a9"
+TestImages = "5e47fb64-e119-507b-a336-dd2b206d9990"
 
 [targets]
-test = ["Test", "Documenter", "IterTools", "LinearAlgebra", "FillArrays", "ComponentArrays"]
+test = ["Test", "Documenter", "IterTools", "LinearAlgebra", "FillArrays", "ComponentArrays", "Images", "ImageQualityIndexes", "TestImages"]
```
Comment (nikopj) [inline Project.toml:56]: this may be a mistake on my part, including Images etc. in [targets]

**Comment by nikopj** [comment]
Some tests in the latest commit are marked as broken for two reasons, to do with F16 #2184 (bug reported in #2190):
- Float16 `conv` is broken for 5D tensors
- Float16 `conv` gradient is broken for 3D/4D/5D tensors, and is also observed in `test/cuda/layers.jl`.

Other than those issues, the ssim loss seems ready to me. I've made the ssim cuda test very verbose to make clear what the issues are exactly, but let me know if its prefered to be a bit more concise.

**Review by johnnychen94 — APPROVED** [review]
Approval for the implementation -- quite straightforward to me.

**Comment by ashwani-rathee** [comment]
@nikopj can we get this going? I had been thinking of more image quality indexes in JuliaImages's ImageQualityIndexes.jl and siphon that through to the DL frameworks

**Comment by nikopj** [comment]
The GPU tests on Julia 1.6 failed because some of my `@test`s used the kwarg `skip=cond`, as some Float16 convolution stuff is still broken as per [https://github.com/FluxML/NNlibCUDA.jl/issues/68](https://github.com/FluxML/NNlibCUDA.jl/issues/68).

My SSIM loss had one failure on Julia 1.8 because of the relative tolerance on a Float16 test being too small:
```bash
f16 cpu-cpu: Test Failed at /var/lib/buildkite-agent/builds/gpuci-2/julialang/flux-dot-jl/test/cuda/losses.jl:45
--
  | Expression: isapprox(loss(f16(x), f16(y)), Float16(loss(x, y)); rtol = 0.1)
  | Evaluated: isapprox(Float16(0.01132), Float16(-0.02692); rtol = 0.1)
```

- Do we avoid using `@test`s kwargs for 1.6 compatibility? should I remove them?
- Is this `rtol` test sensible? I took it from existing F16 tests... though 0.1 seems a bit arbitrary. For reference, the numbers above are small to begin with as they are the SSIM of two random images. The test only fails for `ssim` and not `ssim_loss` or `ssim_loss_fast` because `ssim_loss` is `1-ssim` and as such the relative tolerance is then satisfied. I can perhaps change the `ssim` case to check for F16 matching on a known test image instead of random images.

**Comment by ToucheSir** [comment]
> * Do we avoid using `@test`s kwargs for 1.6 compatibility? should I remove them?

`atol` and `rtol` work in 1.6. What doesn't IIRC is using e.g. `@test broken=true` instead of `@test_broken`.

> * Is this `rtol` test sensible? ...

I'm less worried about the tolerance, but that discrepancy looks pretty concerning?

**Comment by nikopj** [comment]
@ToucheSir It seem's like theres some numerical instability in the Float16 case. Adding some prints into the code (see below), the instability really shows up in the `ssim_map` calculation of `ssim`. My guess is that the C1 and C2 constants were not tuned for F16 arithmetic. Any thoughts @johnnychen94?

```julia
julia> using Flux, Images, TestImages, ImageView

julia> img_x = testimage("monarch_color_256") .|> Gray .|> Float64 |> unsqueeze(dims=3) |> unsqueeze(dims=4);

julia> img_y = testimage("fabio_color_256") .|> Gray .|> Float64 |> unsqueeze(dims=3) |> unsqueeze(dims=4);

julia> Flux.ssim(Float32.(img_x), Float32.(img_y))
C₁ = 0.0001f0
C₂ = 0.0009f0
extrema(μx) = (0.12141457f0, 0.87607676f0)
extrema(μy) = (0.098093934f0, 0.91049606f0)
extrema(μxy) = (0.019157613f0, 0.71647954f0)
extrema(μx²) = (0.014741498f0, 0.7675105f0)
extrema(μy²) = (0.00962242f0, 0.8290031f0)
extrema(σx²) = (2.9206276f-6, 0.103816226f0)
extrema(σy²) = (3.086403f-6, 0.06908727f0)
extrema(σxy) = (-0.035758138f0, 0.04259558f0)
extrema(abs, σx²) = (2.9206276f-6, 0.103816226f0)
extrema(abs, σy²) = (3.086403f-6, 0.06908727f0)
extrema(abs, σxy) = (0.0f0, 0.04259558f0)
extrema((2μxy .+ C₁) .* (2σxy .+ C₂)) = (-0.040532943f0, 0.043345165f0)
extrema(abs, (2μxy .+ C₁) .* (2σxy .+ C₂)) = (3.2075288f-8, 0.043345165f0)
extrema(((μx² .+ μy²) .+ C₁) .* ((σx² .+ σy²) .+ C₂)) = (4.8610345f-5, 0.093692295f0)
extrema(abs, ((μx² .+ μy²) .+ C₁) .* ((σx² .+ σy²) .+ C₂)) = (4.8610345f-5, 0.093692295f0)
extrema(ssim_map) = (-0.8139806f0, 0.9414297f0)
0.13379417f0

julia> Flux.ssim(Float16.(img_x), Float16.(img_y))
C₁ = Float16(0.0001)
C₂ = Float16(0.0009)
extrema(μx) = (Float16(0.12134), Float16(0.876))
extrema(μy) = (Float16(0.09796), Float16(0.909))
extrema(μxy) = (Float16(0.01909), Float16(0.713))
extrema(μx²) = (Float16(0.014725), Float16(0.7676))
extrema(μy²) = (Float16(0.0096), Float16(0.8267))
extrema(σx²) = (Float16(-0.002197), Float16(0.10645))
extrema(σy²) = (Float16(-0.004883), Float16(0.06934))
extrema(σxy) = (Float16(-0.03528), Float16(0.04224))
extrema(abs, σx²) = (Float16(0.0), Float16(0.10645))
extrema(abs, σy²) = (Float16(0.0), Float16(0.06934))
extrema(abs, σxy) = (Float16(0.0), Float16(0.04224))
extrema((2μxy .+ C₁) .* (2σxy .+ C₂)) = (Float16(-0.03915), Float16(0.04926))
extrema(abs, (2μxy .+ C₁) .* (2σxy .+ C₂)) = (Float16(9.0e-7), Float16(0.04926))
extrema(((μx² .+ μy²) .+ C₁) .* ((σx² .+ σy²) .+ C₂)) = (Float16(-0.001884), Float16(0.09894))
extrema(abs, ((μx² .+ μy²) .+ C₁) .* ((σx² .+ σy²) .+ C₂)) = (Float16(2.91e-5), Float16(0.09894))
extrema(ssim_map) = (Float16(-12.11), Float16(6.137))
Float16(0.1942)
```

Plotting the `ssim_map` from each call you can see the difference easily. ex,
```julia
julia> using Plots

julia> p1 = heatmap(map32[:,:,1,1]; cbar=true, title="ssim_map Float32");

julia> p2 = heatmap(map16[:,:,1,1]; cbar=true, title="ssim_map Float16");

julia> p3 = heatmap(clamp.(map16[:,:,1,1], -1, 1); cbar=true, title="clamp(ssim_map, 0, 1) Float16");

julia> plot(p1, p2, p3; layout=(1,3), size=(1200, 400))
```
ssim_stability

**Comment by nikopj** [comment]
I've moved this to its own package [SSIMLoss.jl](https://github.com/nikopj/SSIMLoss.jl) so it can be used with other frameworks more easily. Thanks @ToucheSir @CarloLucibello @johnnychen94 for the help and feedback.

**Pull request closed without merging**